Mas i1788 reconcileapi

martinsumner opened the pull request:
Complete fetch API for PB as well as http.  Add Push API, to allow for discovered differences to be re-queued.

The intention is that this can be used to assist when reconciling a Riak store against an external data store.  Some testing outside of riak_test has been done with dynamodb.  The reconciliation is originally intended to be one-way (e.g. assuming the  non-Riak data store is a read-only copy).

The fetch API is extended to support a different encoding of the fetched object, `internal_aaehash`.  This encoding provides the SegmentId and SegmentHash for this version of the object, to allow that to be indexed within the external db (e.g. to provide merkle trees for comparison).

Testing - https://github.com/basho/riak_test/pull/1356

martinsumner committed:
Support requeue and segment hashes

martinsumner committed:
Add POST to allowed methods

martinsumner committed:
Standardise JSON format for Keys/Clocks

Use same in receiving on riak_kv_wm_queue as when returning from riak_kv_clusteraae_fsm

martinsumner committed:
Add process_post/2

Required to accept post - functions only run on content_type_provided/2 if request was a GET

martinsumner committed:
Add queue lengths to response

Provide a response body to the post

martinsumner committed:
Update API branch

martinsumner committed:
Add pb support for alternative encoding

martinsumner committed:
Add PB support for Push Request

Push will queue for replication a list of Bucket/Key/Clocks.

As part of this change the riak http client has been updated, with types on the fetch function - and further changes have then been required in riak_kv_replrtq_snk to ensure that dialyzer recognises that the correct types are generated in InitClientFun

e.g. otherwise - `507: The call rhc:fetch(HTC::{'rhc',maybe_improper_list(),integer(),maybe_improper_list(),[any()]}, QN::any()) does not have an opaque term of type rhc:rhc() as 1st argument `

martinsumner committed:
Extend supported req IDs for rpbpushreq

martinsumner committed:
Use binary in reponse

martinsumner committed:
Change for consistency with http client

martinsumner committed:
corrections to and alignment between API

ThomasArts commented on rebar.config:
```diff
@@ -56,9 +56,9 @@
     {riak_pipe, {git, "git://github.com/basho/riak_pipe.git", {tag, "riak_kv-3.0.5"}}},
     {riak_dt, {git, "git://github.com/basho/riak_dt.git", {tag, "riak_kv-3.0.0"}}},
-    {riak_api, {git, "git://github.com/basho/riak_api.git", {tag, "riak_kv-3.0.5"}}},
+    {riak_api, {git, "git://github.com/basho/riak_api.git", {branch, "mas-i1788-reconcileapi"}}},
```
A little to-do after update of sub projects to point to new tag

martinsumner replied:
Yes, will pick this up as part of the release cycle

ThomasArts commented on src/riak_kv_pb_object.erl:
```diff
@@ -379,6 +434,38 @@ process_stream(_,_,State) ->
+        {riak_object:bucket(RObj), riak_object:key(RObj), vclock:vclock(),
+            to_fetch}.
+unpack_keyclock_fun(RpbKeysClock) ->
+    case RpbKeysClock#rpbkeysvalue.type of
```
This is code duplication. I wonder if it would not be easier to write 
```suggestion
    Bucket = case RpbKeysClock#rpbkeysvalue.type of
                             undefined -> RpbKeysClock#rpbkeysvalue.bucket;
                             T -> {T, RpbKeysClock#rpbkeysvalue.bucket}
                      end,
       {Bucket, RpbKeysClock#rpbkeysvalue.key,
                 riak_object:decode_vclock(RpbKeysClock#rpbkeysvalue.value),
                 to_fetch}
```

martinsumner replied:
Agreed, will change

ThomasArts commented on src/riak_kv_replrtq_snk.erl:
```diff
@@ -530,18 +529,7 @@ map_peer_to_wi_fun({QueueName, Iteration, PeerInfo}) ->
                                 {certfile, CertfiicateFilename},
```
This variable name has a typo, but nothing you did add.

martinsumner replied:
Will still change though

ThomasArts commented on src/riak_kv_wm_queue.erl:
```diff
@@ -25,10 +25,16 @@
 %% GET /queuename/QueueName
 %%
-%% Will return an object id an object present on queue
+%% Will return an object if an object is present on the queue
```
```suggestion
%% Will return an object if an object is present in the queue
```

ThomasArts commented on src/riak_kv_wm_queue.erl:
```diff
@@ -63,12 +79,12 @@ init(Props) ->
     case riak_kv_wm_utils:get_riak_client(RiakProps, ClientID) of
         {ok, C} ->
-            {true, RD, Ctx#ctx{client = C}};
+            {true, RD, Ctx#ctx{client = C, method=wrq:method(RD)}};
```
This surprises me... should you really update the context with the method type?

Is that because this is used in malformed_request? If so, I would rather get the method from RD in malformed_request instead of counting on this effect to take place.

martinsumner replied:
This may have come about as a copy & paste from https://github.com/basho/riak_kv/blob/develop-3.0/src/riak_kv_wm_object.erl#L222.

I think there is a general problem that I've not been clear about the correct way of working with webmachine, and so this is rather cobbled from other examples from within the riak code.

The same is true with the misuse of malformed request, this is a pattern common elsewhere in riak.

I'm going to look through this agin, and see if it can be better organised.

martinsumner replied:
Fixed

ThomasArts commented on src/riak_kv_wm_queue.erl:
```diff
@@ -77,25 +93,100 @@ service_available(RD, Ctx=#ctx{riak=RiakProps}) ->
     ObjectFormatRaw = wrq:get_qs_value(?Q_OBJECT_FORMAT, "internal", RD),
     ObjectFormat = list_to_atom(ObjectFormatRaw),
-    {false, RD, Ctx#ctx{queuename = QueueName, object_format = ObjectFormat}}.
+    {false, RD, Ctx#ctx{queuename = QueueName, object_format = ObjectFormat}};
```
This is actually not "checking" for malformed ness. If one of the commands crashes, you may get a malformed message... not even sure, but basically you always get a `false` returned, meaning is is well formed.

I would expect that if the QueuName decoding fails or `get_qs_value` does not find the query, that you then return `true`... . Not that `wrq:get_qs_value` may return undefined, in which case the list_to_atom crashes. If you really want an error message  would just crash if the queue is not there. This would cause a 500 internal error instead of a 400 bad request.

martinsumner replied:
Some general work done now to try and return errors in expected error cases

ThomasArts commented on src/riak_kv_wm_queue.erl:
```diff
@@ -77,25 +93,100 @@ service_available(RD, Ctx=#ctx{riak=RiakProps}) ->
+    {false, RD, Ctx#ctx{queuename = QueueName, object_format = ObjectFormat}};
+malformed_request(RD, Ctx) when Ctx#ctx.method =:= 'POST' ->
+    QueueName = malformed_queuename(RD),
+    KeyClockList =
```
Same here... this kind of crashes or always succeeds

martinsumner replied:
Now errors

ThomasArts commented on src/riak_kv_wm_queue.erl:
```diff
@@ -77,25 +93,100 @@ service_available(RD, Ctx=#ctx{riak=RiakProps}) ->
+    B = 
+        case lists:keyfind(<<"bucket">>, 1, BKC) of
+            {<<"bucket">>, Bucket} when is_binary(Bucket) ->
+                case lists:keyfind(<<"bucket-type">>, 1, BKC) of
```
Where is this encoded? How do I know for sure that the encoder uses `bucket-type` as key and not `bucketType` or so?

What I try to say.... it is nice if encoder and decoder are in the same module and if the use of decode here uses the module in which the encoder is defined

martinsumner replied:
This is a won't fix for now.  It is a common problem with the http api vs the pb api in riak - with the pb api there is the riak_pb library shared between clients and server, but no such thing for http.

A generic problem, rather than one we can fix entirely here.

ThomasArts commented on src/riak_kv_wm_queue.erl:
```diff
@@ -77,25 +93,100 @@ service_available(RD, Ctx=#ctx{riak=RiakProps}) ->
+                    false ->
+                        Bucket
+                end;
+            false ->
```
So you allow JSON elements in this JSON Array that do not have "bucket" as key. That's a slippery slope I guess. This is where malformed would check that they are all well formed.

martinsumner replied:
Now any incorrect element in the list will lead to the request being considered malformed

ThomasArts commented on src/riak_kv_wm_queue.erl:
```diff
@@ -77,25 +93,100 @@ service_available(RD, Ctx=#ctx{riak=RiakProps}) ->
+                io_lib:format("No queue ~w", [QueueName])
+        end,
+    {true,
+        wrq:set_resp_body(
```
This is not very REST. One would expect to get the response as `application/json` instead of text. That means, instead of creating the io:format above, build a JSON structure.

martinsumner replied:
Acknowledged, but not fixed.

ThomasArts commented on src/riak_kv_wm_queue.erl:
```diff
@@ -107,9 +198,36 @@ produce_queue_fetch(RD, Ctx) ->
+    ObjBin = encode_riakobject(RObj),
+    TombClockBin = term_to_binary(TombClock),
+    TCL = byte_size(TombClockBin),
+    {<<SuccessMark/binary, IsTombstone/binary,
```
Should this really be part of web machine queue handler? So much detail in building a binary that I can hardly believe this is the right place. Rather have it somewhere in are or so...

But also, would you want this as a binary or rather as a JSON object if you're anyway doing it for having 3rd party apps handle the output.

martinsumner replied:
To  certain extent the use of a binary is locked in here, as this is how it is done for nextgenrepl.  I think the right thing to do for the future is to properly support an alternative to the `internal` and `internal_aaehash` encoding options which is to have a `external_json` alternative.

ThomasArts reviewed (commented):
Looks ok on first sight. Wonder though about the encoding/decoding towards 3rd party applications. I guess the idea is to outward produce JSON instead of binaries in which encoding is hidden... but for efficiency that is an option, of course.

ThomasArts commented on src/riak_kv_wm_queue.erl:
```diff
@@ -77,25 +93,100 @@ service_available(RD, Ctx=#ctx{riak=RiakProps}) ->
+    {struct, [{<<"keys-clocks">>, KCL}]} = mochijson2:decode(ReqBody),
+    lists:reverse(lists:foldl(fun decode_bucketkeyclock/2, [], KCL)).
+
+decode_bucketkeyclock({struct, BKC}, Acc) ->
```
Suggestions, just map over the list and raise an error if <<"key">>, <<"clock">> or other expected fields are missing.

martinsumner replied:
This has now changed so that we check for nay elements being filtered out, due to incorrect encoding, an the whole request will be malformed unless all elements are correct

martinsumner committed:
Refactor following review

martinsumner committed:
Update src/riak_kv_wm_queue.erl

Co-authored-by: Thomas Arts <thomas.arts@quviq.com>

martinsumner committed:
Switch to reject as malformed request if any element of push is malformed

It would be confusing to just filter them out, as other than looking at the logs, it is not clear what has happened.  No obvious reason either, why some should be valid and some not.

martinsumner committed:
Merge branch 'mas-i1788-reconcileapi' of https://github.com/basho/riak_kv into mas-i1788-reconcileapi

martinsumner committed:
Update rebar.config
